### PR TITLE
Add test plan for math array types

### PR DESCRIPTION
**Pull request by nolmoonen**
Adds a test plan for math array types (`marray`) as described in Section 4.14.3 of the SYCL 2020 specification.

### CODE REVIEW

File: test_plans/math_array.asciidoc [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+* `std::uint8_t`
+* `std::int16_t`
+* `std::uint16_t`
+* `std::uint32_t`
+* `std::int64_t`
+* `std::uint64_t`
+
+=== Dimension coverage
+
+The interface tests and memory layout and alignment tests (Sections 2.1 and 2.3 below) are performed using each of the following dimensions for the `NumElements` template variable of `marray`: `1`, `2`, `4`, `8`, `16`, `32`, and `64`, unless stated otherwise.
```
Comment (gmlueck) [inline test_plans/math_array.asciidoc:65]: The `NumElements` template parameter is allowed to be any value of type `size_t`, even ones that are not a power of two.  I think you should check at least one non-power-of-two value (e.g. `5` might be good).  Technically, I think `0` is also allowed.  Would that fall out for free when you write the test, or will it require special handling?  If it falls out for free, it's worth testing.  If `0` requires a lot of extra work, we can save it until after we get the basic tests written.

Comment (Melirius) [reply test_plans/math_array.asciidoc:65]: No, 0 is not allowable, see 4.14.3.

File: test_plans/math_array.asciidoc [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+==== Hidden friend functions
+
+Declare the following `marray<DataT, NumElements>` instances:
+
+* `ma_inc` as the sequence `1`, `2`, `3`, `4`, `5`, etc.
+* `ma_one` as the sequence `0`, `1`, `0`, `1`, `0`, etc.
+* `ma_two` as the sequence `0`, `2`, `0`, `2`, `0`, etc.
+* `ma_alt` as the sequence `1`, `2`, `1`, `2`, `1`, etc.
+
+Below is a table specifying the input for each of the hidden friend functions. Two pairs of inputs are specified: the first pair specifies input with a right-hand side `marray` and a left-hand side with a scalar or another `marray`. The right-hand side `marray` is initialized to be the same constant value as the scalar in the table. The second pair specifies input with a left-hand side of a scalar or a `marray` and a right-hand side with another `marray`. The left-hand side `marray` is initialized to be the same constant value as the scalar in the table. The result column for commutative operator table entries for the second pair are not filled as they are equal to the repective column for the first pair. N.D. denotes "not defined".
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/math_array.asciidoc]: Fixed.

File: test_plans/math_array.asciidoc [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+* `marray<bool, NumElements> operatorOP(const DataT& lhs, const marray& rhs)`, `OP` is `==`, `!=`, `<`, `>`, `\<=`, or `>=`.
+* `marray& operator~(const marray& v)`
+* `marray<bool, NumElements> operator!(const marray& v)`
+
+=== Alias tests
+This section tests Section 4.14.3.2 of the SYCL 2020 specification.
+
+This test is executed for `<elems>` and pairings of `<type>, <storage-type>`. `<elems>` is `2`, `3`, `4`, `8`, and `16`. `<type>` and `<storage-type>` is `char` and `int8_t`, `uchar` and `uint8_t`, `short` and `int16_t`, `ushort` and `uint16_t`, `int` and `int32_t`, `uint` and `uint32_t`, `long` and `int64_t`, `ulong` and `uint64_t`, `float` and `float`, and `bool` and `bool`. In addition, if the device has `aspect::fp64`: `double` and `double`. Furthermore, if the device has `aspect::half`: `half` and `half`.
+
+* Check that `std::is_same_v<m<type><elems>, marray<<storage-type>, <elems>>>`.
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/math_array.asciidoc]: This seems to fix it, thanks.

File: test_plans/math_array.asciidoc [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+
+=== Memory layout and alignment tests
+This section tests Section 4.14.3.3 of the SYCL 2020 specification.
+
+* Check that `alignof(marray<DataT, NumElements>)` is equal to `alignof(std::array<DataT, NumElements>)`.
+* Check that `sizeof(marray<DataT, NumElements>)` is equal to `sizeof(std::array<DataT, NumElements>)`.
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/math_array.asciidoc:449]: Good point, I have added this test.

File: test_plans/math_array.asciidoc (hunk 1 above)
Comment (nolmoonen) [reply test_plans/math_array.asciidoc]: Thanks for the suggestion, I have added a non-power-of-two value.

I agree with Melirius, the spec states:
> The number of elements parameter, `NumElements`, is a positive value of the `std::size_t` type.

In the case that we are interpreting it incorrectly, `0` can be added without much effort (some tests should slightly be modified when `NumElements` is `0`, such as the iterator test).

File: test_plans/math_array.asciidoc [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+* `marray<RET, NumElements> operatorOP(const DataT& lhs, const marray& rhs)`, `OP` is `&&` or `||`.
+* `marray operatorOP(const DataT& lhs, const marray& rhs)`, `OP` is `<<` or `>>`.
+* `marray<bool, NumElements> operatorOP(const DataT& lhs, const marray& rhs)`, `OP` is `==`, `!=`, `<`, `>`, `\<=`, or `>=`.
+* `marray& operator~(const marray& v)`
+* `marray<bool, NumElements> operator!(const marray& v)`
+
+=== Alias tests
+This section tests Section 4.14.3.2 of the SYCL 2020 specification.
+
+This test is executed for `<elems>` and pairings of `<type>, <storage-type>`. `<elems>` is `2`, `3`, `4`, `8`, and `16`. `<type>` and `<storage-type>` is `char` and `int8_t`, `uchar` and `uint8_t`, `short` and `int16_t`, `ushort` and `uint16_t`, `int` and `int32_t`, `uint` and `uint32_t`, `long` and `int64_t`, `ulong` and `uint64_t`, `float` and `float`, and `bool` and `bool`. In addition, if the device has `aspect::fp64`: `double` and `double`. Furthermore, if the device has `aspect::half`: `half` and `half`.
```
Comment (keryell) [inline test_plans/math_array.asciidoc:441]:
```suggestion
This test is executed for `<elems>` and pairings of `<type>, <storage-type>`. `<elems>` is `2`, `3`, `4`, `8`, and `16`. `<type>` and `<storage-type>` is `char` and `int8_t`, `uchar` and `uint8_t`, `short` and `int16_t`, `ushort` and `uint16_t`, `int` and `int32_t`, `uint` and `uint32_t`, `long` and `int64_t`, `ulong` and `uint64_t`, `float` and `bool`. In addition, if the device has `aspect::fp64`: `double`. Furthermore, if the device has `aspect::half`: `half`.
```
Please use short lines in the document.

**Review by keryell — CHANGES_REQUESTED** [review]

File: test_plans/math_array.asciidoc [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+=== Type coverage
+
+The interface tests and memory layout and alignment tests (Sections 2.1 and 2.3 below) are performed using each of the following types as the underlying type the `DataT` template variable of `marray`, unless stated otherwise.
+
+In regular mode:
+
+* `char`
+* `int`
+* `float`
+* `std::byte`
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/math_array.asciidoc]: I removed the type.

File: test_plans/math_array.asciidoc [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+
+* `half`
+
+In full conformance mode:
+
+* `signed char`
+* `unsigned char`
+* `short`
+* `unsigned short`
+* `unsigned int`
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/math_array.asciidoc:49]: The full conformance mode is intended to also include the types for regular mode. I have updated
> In full conformance mode:

to
> In full conformance mode, the following types are additionally tested:

File: test_plans/math_array.asciidoc [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+* `unsigned long`
+* `long long`
+* `unsigned long long`
+* `bool`
+* `std::uint8_t`
+* `std::int16_t`
+* `std::uint16_t`
+* `std::uint32_t`
+* `std::int64_t`
+* `std::uint64_t`
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/math_array.asciidoc:60]: See above comment.

File: test_plans/math_array.asciidoc [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+
+=== Dimension coverage
+
+The interface tests and memory layout and alignment tests (Sections 2.1 and 2.3 below) are performed using each of the following dimensions for the `NumElements` template variable of `marray`: `1`, `2`, `4`, `5`, `8`, `16`, `32`, and `64`, unless stated otherwise.
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/math_array.asciidoc]: I see no downside in adding it, so I have done that.

File: test_plans/math_array.asciidoc [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+== Tests
+
+=== Interface tests
+
+This section tests Section 4.14.3.1 of the SYCL 2020 specification.
+
+==== Constructors
+
+* Call constructor `marray()` to initialize an instance `ma`. Check that all the values of `ma` are equal to the default-constructed elements `DataT`.
+* Call constructor `explicit constexpr marray(const DataT& arg)` with `DataT(1)` to initialize a `constexpr` instance `ma`. Check using `operator[]` that all values of `ma` are equal to `DataT(1)`.
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/math_array.asciidoc]: I had not intended this, but I will now and have added this clarification.

File: test_plans/math_array.asciidoc (hunk 5 above)
Comment (nolmoonen) [reply test_plans/math_array.asciidoc]: Fixed.

**Comment by nolmoonen** [comment]
I have changed Section 2.1.3 "Hidden friend functions" of the test plan, due to it containing an error. The test used very specific input and ouput, which does not work when `DataT` is `bool`. Instead, I opted for a simpler approach that performs the operations using some predefined instances. The same operations are performed using `std::valarray`, with which the result is checked. This has the added benefit that will be easier to verify that the test does what it is supposed to do. Furthermore, this gets rid of the awkward edge cases for the unary operators.

@gmlueck Could you perhaps also see if you agree with these changes?

File: test_plans/math_array.asciidoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (keryell) [reply test_plans/math_array.asciidoc:49]: Ah I see. Perhaps it would be clearer to have an extensive description to avoid intensive reflection and repeat in the second part what is said in the first part to avoid forgetting some types.

[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/math_array.asciidoc:51]: Alright, I have applied your suggestion.

**Comment by gmlueck** [comment]
> Instead, I opted for a simpler approach that performs the operations using some predefined instances. The same operations are performed using std::valarray, with which the result is checked.

This does seem simpler.  The only problem I see is that `std::valarray` doesn't seem to provide `operator++` or `operator--` (at least according to [cppreference](https://en.cppreference.com/w/cpp/numeric/valarray)).  Does anyone know why they are missing?

**Comment by keryell** [comment]
> This does seem simpler. The only problem I see is that `std::valarray` doesn't seem to provide `operator++` or `operator--` (at least according to [cppreference](https://en.cppreference.com/w/cpp/numeric/valarray)). Does anyone know why they are missing?

It is not the only problem.
But here we can just inherit from `std::valarray` and add these 2 operators.

**Review by keryell — APPROVED** [review]
Thanks.
